### ISSUE TITLE
Pyright does not narrow TypeGuard in magic methods

### ISSUE THREAD

**Issue by kumaraditya303** [microsoft/pyright#2588]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
It does not handles typeguard.

**Expected behavior**
Narrow type to int as per typeguard.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from typing import TypeGuard


class Foo:
    def __contains__(self, item: object) -> TypeGuard[int]:
        return isinstance(item, int)


f = Foo()
for i in [1, 2, "3"]:
    if i in f:
        reveal_type(i)
        print(i)

```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

```console
pyright 1.1.188

```

**Comment by erictraut** [comment]
This isn't a supported type guard pattern. For a complete list of supported type guard patterns in pyright, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards).

Each new type guard pattern requires case-specific code that can slow down overall type analysis, so we're relatively conservative about adding support for new patterns unless we have a strong signal from users that it's an important (common-place) pattern. This sample strikes me as a very odd use case.

For this particular type guard pattern, it's unlikely that we'd ever add support for the following reasons:

1. This isn't consistent with how `TypeGuard` works. If you read PEP 647, you'll note that user-defined type guards accept a value and returns a bool indicating whether that value is a particular type. Those semantics are very different from `__contains__`, which accepts a value and returns a bool indicating whether the object contains the value (by whatever definition the class wants to ascribe to "contains"). It wouldn't make sense to mix up the semantics of these two constructs.

2. The implementation of the `in` operator is much more complex than simply looking up a `__contains__` method. It also potentially involves repeated calls to `__iter__` and/or `__getitem__`. Pyright doesn't fully emulate the internal implementation of CPython's `in` implementation, and it would be complex to do so.

**Comment by kumaraditya303** [comment]
> The implementation of the in operator is much more complex than simply looking up a __contains__ method. It also potentially involves repeated calls to __iter__ and/or __getitem__. Pyright doesn't fully emulate the internal implementation of CPython's in implementation, and it would be complex to do so.

This is very common scenario use case in python , pyright can check if class has `__contains__` if not then fallback but just by supporting `__contains__` would be a huge benefit. Can you convert this issue to feature request to add it ?

**Comment by erictraut** [comment]
As author of the TypeGuard PEP, I'm telling you that this would not be an appropriate use of the mechanism. See my answer above for details.

If you need help finding a solution to a particular problem, please post to the discussion, and someone will attempt to suggest a solution. This isn't the right solution to the problem you're trying to solve.
